src/witan/app/password_reset.clj

```diff
@@ -0,0 +1,46 @@
+(ns witan.app.password-reset
+  (:require [witan.app.util :refer [password-reset-token]]
+            [witan.app.email :refer [send-password-reset!]]
+            [witan.app.config :refer [exec]]
+            [witan.app.user :as usr]
+            [qbits.hayt :as hayt]
+            [clojure.tools.logging :as log]
+            [clojure.core.async :as async]))
+
+(defn begin-reset-password!
+  "Starts the password reset process by issuing a token and storing"
+  [username]
+  (let [user-str (str \[ username \])]
+    (if (not-empty (usr/retrieve-user-by-username username))
+      ;; use an old token if we find one
+      (let [{:keys [password_reset_token]} (first (exec (hayt/select :password_reset_tokens (hayt/where {:username username}))))
+            token (or password_reset_token (password-reset-token))]
+        (log/debug "Generating a password reset email for" user-str)
+        (async/go
+          (try
+            (if-let [{:keys [message]} (send-password-reset! username token)]
```

Comment:
It is slow, yes. Also, with regards to feeding back to the user, we don't care if it succeeds or fails - we always return a 200 for a reset request.